### PR TITLE
Feat/filter interaction

### PR DESCRIPTION
**Pull request by purpose233**
[[中文版模板 / Chinese template](https://github.com/VisActor/vgrammar/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Chore
- [ ] Release
- [ ] Other (about what?)

### 💡 Background and solution

Support filter & popover interactions.
TODO: preview component interactions and tooltip & crosshair component haven't been removed yet, in case the interaction 
 implementation effect the usage of vchart

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vgrammar-core/src/interactions/filter.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  view: IView;
+  _data?: IData;
+  _marks?: IMark[];
+  _filterValue: any;
+  _dataFilter: IDataFilter;
+  handleFilter: (event?: InteractionEvent) => void;
+}
+
+export class FilterMixin {
+  _filterData(
```
Comment (xile611) [inline packages/vgrammar-core/src/interactions/filter.ts:16]: 这个逻辑在 `comonent` 中还有一份？
不是要删掉吗？ 似乎没看到

File: docs/dev-demos/src/api/brush.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  const brushData = view.data().id('brushData').values(originData);
+
+  const colorScale = view.scale('ordinal').domain({ data: data, field: 'type' }).range(category10);
+  const rightLegend = view
+    .legend(view.rootMark)
+    .id('rightLegend')
+    .scale(colorScale)
+    // .target(markData, 'type')
+    .encode({ x: 340, y: 120 });
+  view.interaction('legend-filter', {
```
Comment (xile611) [inline docs/dev-demos/src/api/brush.ts:38]: 按照之前讨论的，不是说`legend-filter` 类的交互，不用显示的调用
默认注册吗？

File: packages/vgrammar-core/src/interactions/tooltip.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { AABBBounds, throttle } from '@visactor/vutils';
```
Comment (xile611) [inline packages/vgrammar-core/src/interactions/tooltip.ts:1]: component 中的tooltip 不删掉吗？

File: packages/vgrammar-core/src/interactions/filter.ts (hunk 1 above)
Comment (purpose233) [reply packages/vgrammar-core/src/interactions/filter.ts:16]: Fixed.

File: docs/dev-demos/src/api/brush.ts (hunk 2 above)
Comment (purpose233) [reply docs/dev-demos/src/api/brush.ts:37]: Fixed.

File: packages/vgrammar-core/src/interactions/tooltip.ts (hunk 3 above)
Comment (purpose233) [reply packages/vgrammar-core/src/interactions/tooltip.ts:1]: Fixed.

File: packages/vgrammar-core/src/interactions/brush-filter.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        this._filterValue = filterValue;
+        this.handleFilter();
+      }
+    }
+
+    this.dispatchEvent(options, elements);
+  };
+}
+
+mixin(BrushFilter, FilterMixin);
```
Comment (xile611) [inline packages/vgrammar-core/src/interactions/brush-filter.ts:80]: 这种副作用相关代码统一挪到注册函数中，不然会被tree-shaking 删掉的，因为我们配置了无副作用

Comment (purpose233) [reply packages/vgrammar-core/src/interactions/brush-filter.ts]: Fixed.

File: packages/vgrammar-core/src/interactions/filter.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  extends Pick<FilterMixin, '_data' | '_marks' | '_filterValue' | '_dataFilter' | 'handleFilter' | '_filterData'>,
+    BaseInteraction {}
+
+export abstract class Filter extends BaseInteraction {
+  static defaultOptions: Omit<DataFilterOptions, 'target'> = {};
+  options: DataFilterOptions;
+
+  constructor(view: IView, options?: DataFilterOptions) {
+    super(view);
+    this._marks = view
```
[the remark this answers is not in the document]
Comment (purpose233) [reply packages/vgrammar-core/src/interactions/filter.ts]: Fixed.

File: packages/vgrammar-core/src/interactions/legend-filter.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import { LegendEvent } from '@visactor/vrender-components';
+import type { DataFilterOptions, IComponent, ILegend, IView, InteractionEvent } from '../types';
+import { DataFilterRank, GrammarMarkType } from '../graph';
+import { isString } from '@visactor/vutils';
+import { Filter } from './filter';
+
+export class LegendFilter extends Filter {
```
[the remark this answers is not in the document]
Comment (purpose233) [reply packages/vgrammar-core/src/interactions/legend-filter.ts:7]: 主要是想着有个 Filter 类方便用户自己继承，要用户写 mixin 来使用 filter 感觉不太友好

**Review by xile611 — APPROVED** [review]

**Pull request merged**